### PR TITLE
Tests v2

### PR DESCRIPTION
**Pull request by XapaJIaMnu**
This should address your comments from the previous pull request

### CODE REVIEW

**Comment by XapaJIaMnu** [comment]
Added CTests integration, consistent naming across tests and so long.
One caveat. The tests will have different matrices generated depending on whether the test was run by itself (by doing `make test` or `./tests "Multiply AVX2 8bit"`). A solution would be to set the random seed before every test function. Do you agree with that?

Cheers,

Nick

**Comment by XapaJIaMnu** [comment]
Also just realised, when compiled on a different hardware depending on the CPU type the RNG will be in a different state because some tests would be skipped.

File: test/multiply_test.cc [hunk 1]
```diff
@@ -244,24 +307,37 @@ template <class Integer> void SlowRefInt(const Integer *A, const Integer *B, flo
...
+    CHECK_MESSAGE(fabs(float_diff) <= float_tolerance, test_info << "Inaccurate compared to float reference at " << i << ' ' << float_ref[i] << ' ' << int_test[i]);
     int_sum += int_diff * int_diff;
     float_sum += float_diff * float_diff;
   }
-  std::cout << "Float MSE = " << sqrt(float_sum / size) << "\tInt MSE = " << sqrt(int_sum / size) << std::endl;
+  CHECK_MESSAGE(fabs(sqrt(float_sum / size)) <= MSE_float_tolerance, test_info << "Float MSE = " << sqrt(float_sum / size));
+  CHECK_MESSAGE(fabs(sqrt(int_sum / size)) <= MSE_int_tolerance, test_info << "Int MSE = " << sqrt(int_sum / size));
+  //We want to see a test failure when we fix some of the accumulation errors, so that we also
+  //fix our tests after we fix the issue. The easiest way to test this is through the MSE
+  if (MSE_float_tolerance > 0.1) {
```
Comment (kpu) [inline test/multiply_test.cc]: I think this is weird.

File: test/multiply_test.cc [hunk 2]
```diff
@@ -244,24 +307,37 @@ template <class Integer> void SlowRefInt(const Integer *A, const Integer *B, flo
...
+  }
+
+  if (MSE_int_tolerance > 0.03) {
+    CHECK_FALSE_MESSAGE(fabs(sqrt(int_sum / size)) <= MSE_int_tolerance/2, test_info << "Int MSE = " << sqrt(int_sum / size) << 
+      " much better than the expected: " << MSE_int_tolerance);
+  }
 }
 
-template <class Routine> void TestMultiply(Index A_rows, Index width, Index B_cols) {
+template <class Routine> void TestMultiply(Index A_rows, Index width, Index B_cols,
```
Comment (kpu) [inline test/multiply_test.cc]: At some point we start developing a struct to pass tolerances, but I see why you did this for now

File: test/multiply_test.cc [hunk 3]
```diff
@@ -292,60 +368,80 @@ template <class Routine> void TestMultiply(Index A_rows, Index width, Index B_co
...
-    TestSelectColumnsB<SSE2_16bit>(256, 256);
-    TestMax<__m128>();
-    TestMaxAbsolute<SSE2_MaxAbsolute>();
-/*    if (kCPU >= CPU_AVX2) {
-      TestMaxAbsolute<AVX2_MaxAbsolute>();
-    }*/
+} // namespace intgemm
+
+int main(int argc, char ** argv) {
+    std::srand(45678); //If the random seed is not crucial we can forego the main alltogether
```
Comment (kpu) [inline test/multiply_test.cc]: If we're concerned about random initialization being consistent on different CPUs, why not use the C++11 random stuff and seed each time the random number generator is called?

File: test/multiply_test.cc [hunk 4]
```diff
@@ -292,60 +368,80 @@ template <class Routine> void TestMultiply(Index A_rows, Index width, Index B_co
...
+    int result = Catch::Session().run( argc, argv );
+    return result;
```
Comment (kpu) [inline test/multiply_test.cc]: return Catch::Session().run(argc, argv);

**Pull request merged**